src/test/kotlin/com/jetbrains/snakecharm/fixtures/SnakemakeResolveTestCase.kt

```diff
@@ -0,0 +1,134 @@
+            // TODO: check and fix work with single letter identifiers
+            val document = PsiDocumentManager.getInstance(psiFile.project).getDocument(psiFile)!!
+            var offset = -1
+            for (i in 1 until document.lineCount) {
+                val lineStart = document.getLineStartOffset(i)
+                val lineEnd = document.getLineEndOffset(i)
+                val index = document.charsSequence.subSequence(lineStart, lineEnd).toString().indexOf("<ref>")
+                if (index > 0) {
+                    offset = document.getLineStartOffset(i - 1) + index
+                }
+            }
+            assertTrue("<ref> in test file not found", offset >= 0)
+            return offset
+        }
+
+        fun findReferenceByMarker(psiFile: PsiFile?): PsiReference? {
+            val ref = psiFile?.findReferenceAt(findMarkerOffset(psiFile))
+            assertNotNull("No reference found at <ref> position", ref)
+            return ref
+        }
+
+        protected fun assertIsBuiltin(element: PsiElement?) {
+            assertNotNull(element)
+            assertTrue(PyBuiltinCache.getInstance(element).isBuiltin(element))
+        }
+    }
+
+    protected fun configureByFile(@TestDataFile filePath: String): PsiReference? {
+        val testDataRoot = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(getTestDataPath().toFile())
+        val file = testDataRoot!!.findFileByRelativePath(filePath)
+        assertNotNull(file)
+
+        var fileText: String
+        try {
+            fileText = StringUtil.convertLineSeparators(VfsUtil.loadText(file!!))
+        } catch (e: IOException) {
+            throw RuntimeException(e)
+        }
+
+        val offset = fileText.indexOf(MARKER)
+        assertTrue(offset >= 0)
+        fileText = fileText.substring(0, offset) + fileText.substring(offset + MARKER.length)
+        fixture?.configureByText(File(filePath).name, fileText)
+        return fixture?.file?.findReferenceAt(offset)
+    }
+
+    protected abstract fun doResolve(): PsiElement?
+
+    protected fun <T : PsiElement> assertResolvesTo(langLevel: LanguageLevel, aClass: Class<T>, name: String): T {
+        val result = Ref<T>()
+
+        runWithLanguageLevel(
+                langLevel,
+                Runnable { result.set(assertResolvesTo(aClass, name, null)) }
+        )
+
+        return result.get()
+    }
+
+    protected fun <T : PsiElement> assertResolvesTo(aClass: Class<T>, name: String): T {
+        return assertResolvesTo(aClass, name, null)
+    }
+
+    protected fun <T : PsiElement> assertResolvesTo(aClass: Class<T>,
+                                                    name: String,
+                                                    containingFilePath: String?): T {
+        val element: PsiElement?
+        try {
+            element = doResolve()
+        } catch (e: Exception) {
+            throw RuntimeException(e)
+        }
```

Comment:
In Kotlin all exceptions are checked, we don't need to re-throw exception here, see https://kotlinlang.org/docs/reference/java-interop.html#checked-exceptions. Please clean up the code, it also could be converted to expression body e.g. `fun foo() = boo(...)`

Reply:
Done